x-pack/solutions/observability/plugins/observability/server/services/related_dashboards_client.ts

```diff
@@ -124,9 +176,19 @@ export class RelatedDashboardsClient {
     const {
       result: { hits },
     } = dashboards;
-    hits.forEach((dashboard: Dashboard) => {
+    for (const dashboard of hits) {
+      for (const panel of dashboard.attributes.panels) {
+        if (
+          isDashboardPanel(panel) &&
+          isSuggestedDashboardsValidPanelType(panel.type) &&
+          (isEmpty(panel.panelConfig) || !panel.panelConfig.attributes)
+        ) {
+          await this.referencedPanelManager.fetchReferencedPanel({ dashboard, panel });
```

Comment:
I think there's still another optimization to be taken here.

Currently, every time we find a panel reference we stop and fetch the panel. 

I'd like to see a parsing step first where we stop, look through all the dashboards we've fetched, find all the panel references that need to be fetched, fetch them all in one bulk request, and then inject them back into the dashboards so we can evaluate relevance. What's your thoughts about that approach?

Reply:
I gave it a try and went ahead with the suggested optimization [here](https://github.com/elastic/kibana/pull/228811/commits/84624387d67d1cd0ed3802d341c7455cb94f41a4).